File: 126 127 Word Ladder.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <stack>
#include <unordered_set>
#include <unordered_map>
/*
126 Word Ladder
127 Word Ladder II
https://oj.leetcode.com/problems/word-ladder/
https://oj.leetcode.com/problems/word-ladder-ii/

Input sample:
hit
cog
hot dot dog lot log

   hit => start
    |
   hot => entries   \
  /   \
dot    lot           tree                       
|       |
dog    log => exits /
  \   /
   cog => ended
*/
typedef char** MAT;
typedef std::vector<std::string> STRLINE;
typedef std::vector<int> INTLINE;
typedef std::unordered_map<int,INTLINE> ROW;//string index : upper level indexes
typedef std::unordered_set<int> SET;
typedef std::vector<ROW> TREE;

bool isLinked(const std::string& tag, const std::string& src){
    int diff = 0;
    auto i = tag.begin();
    auto j = src.begin();
    for(;i!=tag.end() && j!=src.end() && diff<2;i++,j++)
        if(*i!=*j) diff++;
    return diff == 1;
}
/*
init exit set
init graph
init entry row
*/
void init(const std::string& start, const std::string& ended, const STRLINE& dict,
ROW& entries, SET& exits, MAT& graph, int& N){
    N = (int)dict.size();
    graph = (char**)malloc(N*sizeof(char*));
    
    for(int i=0;i!=N;i++){
        if(isLinked(start,dict[i])){
            INTLINE vec(1,-1); // start std::string index is -1
            entries.insert(std::pair<int,INTLINE>(i,vec));
        }
        if(isLinked(ended,dict[i])) exits.insert(i);
        
        graph[i] = (char*)calloc(N,sizeof(char));
        for(int j=0;j!=N;j++){     
            if(j==i) continue;
            if(isLinked(dict[i],dict[j])) graph[i][j] = 1;
        }
    }
#if OUTPUT_FILE
	ofstream out("../out.txt");
	for(int i=0;i!=N;i++){
		for(int j=0;j!=N;j++){
			char c = graph[i][j];
			if(c==1) out<<"1 ";
			else out<<"0 ";
		}
		out<<"\n";
	}
	out.close();
#endif
}
/*
Find if a key exists in upper tree rows
*/
bool isKeyIn(TREE& tree, int& tag){
    for(auto i = tree.begin();i!=tree.end();i++){
        if(i->find(tag)!=i->end()) return true;
    }
    return false;
}
/*
*/
bool iterate(ROW& entries, SET& exits, MAT& graph, const int& N, TREE& tree){
    int solCtr = 0;
	if(entries.empty()) return 0;
    tree.push_back(entries);
    for(std::size_t level = 0;level<N-entries.size();level++){
        ROW& curr = tree[level];
        ROW next;
        for(auto i = curr.begin();i!=curr.end();i++){
            int x = i->first;
            if(exits.find(x)!= exits.end()) return 1;
            
            // find linked element of x in graph[x]
            for(auto k = 0;k<N;k++)
                if(graph[x][k]==1)
                    if(!isKeyIn(tree,k))
						next[k].push_back(i->first);           
        }
        if(next.empty()) return 0;
		tree.push_back(next);
    }
    return 0;
}
/*
*/
void outputRec(TREE& tree,
std::vector<std::stack<int> >& res, std::stack<int>& buf, int level, int node){
    auto in  = tree[level].find(node); // must found
    buf.push(in->first);
    if(level==0){
        res.push_back(buf); return;
    }
    
    INTLINE& from = in->second;
    for(auto j = from.begin();j!=from.end();j++){
        std::stack<int> newbuf(buf);
        outputRec(tree,res,newbuf,level-1,*j);
    }
}
/*
Notice exits are not necessary equal to bottom level of tree
*/
void output(const std::string& start, const std::string& ended, const STRLINE& dict,
SET& exits, TREE& tree){
    std::vector<std::stack<int> > res;
    std::size_t N = tree.size();
    for(auto i = exits.begin();i!=exits.end();i++){
        auto in  = tree[N-1].find(*i);
        if(in != tree[N-1].end()){
            INTLINE& from = in->second;
            for(auto j = from.begin();j!=from.end();j++){
                std::stack<int> buf;
                buf.push(*i);
                outputRec(tree,res,buf,N-2,*j);
            }
        }
    }
    
    for(auto i = res.begin();i!=res.end();i++){
        std::stack<int>& buf = *i;
        std::cout<<start<<" ";
        while(!buf.empty()){
            std::cout<<dict[buf.top()]<<" ";
            buf.pop();
        }
        std::cout<<ended<<"\n";
    }
    std::cout<<"\n";
}

int main(int argc, char* argv[]){
    std::ifstream f(argv[1]);	
	std::string start, ended, dictLine,t;
	STRLINE dict;
    ROW entries;
	SET exits;
	MAT graph;
    int N = 0;
	TREE tree;
    
	getline(f, start);
	getline(f, ended);
	getline(f, dictLine);
	
	std::stringstream iss(dictLine);
	while(iss>>t) dict.push_back(t);
	f.close();
	
	init(start, ended, dict, entries, exits, graph,N);
	if(iterate(entries, exits, graph,N, tree))
		output(start, ended, dict, exits,tree);
    return 0;
}

File: 126 127 Word Ladder.py
'''
126 Word Ladder
127 Word Ladder II

Double-sided BFS
Try to find the shorted solutions

Input sample:
hit
cog
hot dot dog lot log

   hit => start
    |
   hot => entries   \
  /   \
dot    lot           tree                       
|       |
dog    log => exits /
  \   /
   cog => ended
'''

pool = []
EMPTY = '-1'

class node:
    '''
    Assume one node has only one daddy
    '''
    def __init__(self,name):
        self.name = name
        self.dady = None
    
def isLinked(src, tag):
    '''
    @src: str
    @tag: str
    '''
    diff = 0
    i = 0 
    j = 0
    while i < len(tag) and j < len(src) and diff < 2:
        if tag[i] != src[j]: diff += 1
        i += 1
        j += 1
    return diff == 1

def hasCommon(upper,lower):
    global pool
    '''
    @upper: list[node]
    @lower: list[node]
    '''
    found = False
    for u in upper:
        for l in lower:
            if isLinked(u.name,l.name):
                seq = [u,l]
                p = u.dady
                while p is not None:
                    seq.insert(0,p)
                    p = p.dady
                p = l.dady
                while p is not None:
                    seq.append(p)
                    p = p.dady
                pool.append(seq)
                found = True
    return found
    
def iterate(remains,upper,lower):
    '''
    @remains: list[str]
    @upper: list[node]
    @lower: list[node]
    '''
    global EMPTY
    if hasCommon(upper,lower): return
    
    newUpper = []
    newLower = []
    
    for i,r in enumerate(remains):
        linked = False
        if r != EMPTY:
            for u in upper:
                if isLinked(r,u.name):
                    n = node(r)
                    n.dady = u
                    newUpper.append(n)
                    linked = True
                    break
            if linked == True: continue
            for l in lower:
                if isLinked(r,l.name):
                    n = node(r)
                    n.dady = l
                    newLower.append(n)
                    linked = True
                    break    
        if linked: remains[i] = EMPTY
        
    if len(newUpper) == 0 and len(newLower) == 0:
        return
    elif len(newUpper) == 0:
        del lower[:]
        iterate(remains,upper,newLower)
    elif len(newLower) == 0:
        del upper[:]
        iterate(remains,newUpper,lower)
    else:
        del upper[:]
        del lower[:]
        iterate(remains,newUpper,newLower)

def main():
    file = open('small.in','r')
    cases = file.readlines()
    lineId = 0
    caseId = 0
    while lineId < len(cases):
        caseId += 1
        
        del pool[:]
        
        s = cases[lineId].strip()
        start = node(s)
        
        s = cases[lineId+1].strip()
        ended = node(s)
        
        remains = cases[lineId+2].strip().split(' ')
        upper = [start]
        lower = [ended]
        
        iterate(remains,upper,lower)
        
        print 'Case #%d:' % (caseId)
        if len(pool) == 0:
            print 'IMPOSSIBLE' 
        else:
            for seq in pool:
                for s in seq:
                    print s.name,
                print
                
        lineId += 3
    file.close()

if __name__ == '__main__':
    main()
